[WGSL] Implement local var statement parsing

djg opened the pull request:
#### 250024bc727642812cf7e5067006871d238ac4fc
<pre>
[WGSL] Implement local var statement parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=244760">https://bugs.webkit.org/show_bug.cgi?id=244760</a>
rdar://problem/99521231

Reviewed by Myles C. Maxfield.

Make use of existing VariableDecl to implement VariableStatement nodes
representing local variable declaration statements beginning with &apos;var&apos; keyword.

* Source/WebGPU/WGSL/AST/Statement.h:
Add Variable to Statement::Kind
(WGSL::AST::Statement::isVariable const):
Predicate for new VariableStatement node
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseVariableDecl):
(WGSL::Parser&lt;Lexer&gt;::parseVariableDeclWithAttributes):
Only global variables can have attributes. Split parsing into parseVariable that
passes an empty list of Attributes for local variable decls and
parseVariableDeclWithAttributes for global variable decls.
(WGSL::Parser&lt;Lexer&gt;::parseStatement):
Handle parsing of variable statements beginning with &apos;var&apos; keyword.
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingLocalVariable]):
Simple test of a function with a local variable.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
Add VariableStatement.h

Canonical link: <a href="https://commits.webkit.org/254140@main">https://commits.webkit.org/254140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7a4f9ac549df8db2d97518d20cc6afe87439347

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97350 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30751 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26653 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80305 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92036 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93783 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24749 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74842 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79667 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28395 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28474 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14692 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2909 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31541 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/37588 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/30482 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33892 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/AST/Statements/VariableStatement.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace WGSL::AST {
+
+class VariableStatement final : public Statement {
```
Nit: VariableDeclarationStatement?

djg replied:
My intention was to match the names of the grammar in the WGSL spec. https://www.w3.org/TR/WGSL/#syntax-variable_statement

litherum commented on Source/WebGPU/WGSL/AST/Statements/VariableStatement.h:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    Kind kind() const override { return Kind::Variable; }
+    Decl& decl() { return m_decl.get(); }
```
This is totally a style thing, but WebKit doesn't usually use abbreviations. We usually spell stuff out to make it easier to read.

djg replied:
Should `Decl` nodes be `Declaration` too? The spec calls it "variable_decl" which I'd like to stylise as `VariableDecl` https://www.w3.org/TR/WGSL/#syntax-variable_decl

litherum replied:
I vote yes, but it's up to you.

litherum commented on Source/WebGPU/WGSL/AST/Statements/VariableStatement.h:
```diff
@@ -0,0 +1,53 @@
+    Decl& decl() { return m_decl.get(); }
+
+private:
+    UniqueRef<Decl> m_decl;
```
What's the point of this class? Why not just put an enum inside `Decl`?

djg replied:
I thought about this too, but it's an adapter from `Statement` to `Decl`. A `CompoundStatement` holds `Statements` and one of those statement kinds is `VariableStatement` that is going hold a `Decl` for `var`, `let`, `const` etc.

My current position is to implement the AST following the WGSL grammar in the spec. We can "optimise" the AST later. What do you think?

litherum replied:
Sounds good to me.

litherum commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -349,7 +358,11 @@ Expected<AST::VariableDecl, Error> Parser<Lexer>::parseVariableDecl(AST::Attribu
     }
 
     std::unique_ptr<AST::Expression> maybeInitializer = nullptr;
-    // FIXME: initializer
```
oof

djg committed:
[WGSL] Implement local var statement parsing
https://bugs.webkit.org/show_bug.cgi?id=244760
rdar://problem/99521231

Reviewed by Myles C. Maxfield.

Make use of existing VariableDecl to implement VariableStatement nodes
representing local variable declaration statements beginning with 'var' keyword.

* Source/WebGPU/WGSL/AST/Statement.h:
Add Variable to Statement::Kind
(WGSL::AST::Statement::isVariable const):
Predicate for new VariableStatement node
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseVariableDecl):
(WGSL::Parser<Lexer>::parseVariableDeclWithAttributes):
Only global variables can have attributes. Split parsing into parseVariable that
passes an empty list of Attributes for local variable decls and
parseVariableDeclWithAttributes for global variable decls.
(WGSL::Parser<Lexer>::parseStatement):
Handle parsing of variable statements beginning with 'var' keyword.
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingLocalVariable]):
Simple test of a function with a local variable.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
Add VariableStatement.h

Canonical link: https://commits.webkit.org/254140@main
